Sources/SwiftProtobuf/ProtobufTextDecoding.swift

```diff
@@ -0,0 +1,1229 @@
+                        throw ProtobufDecodingError.malformedJSON
+                    }
+                    break
+                case .endArray:
+                    return true
+                default:
+                    throw ProtobufDecodingError.malformedJSON
+                }
+            }
+        }
+    }
+    
+    mutating func decodePackedField<S: ProtobufTypeProperties>(fieldType: S.Type, value: inout [S.BaseType]) throws -> Bool {
+        return try decodeRepeatedField(fieldType: fieldType, value: &value)
+    }
+    
+    mutating func decodeSingularMessageField<M: ProtobufMessage>(fieldType: M.Type, value: inout M?) throws -> Bool {
+        var m = value ?? M()
+        var subDecoder = ProtobufTextDecoder(scanner: scanner)
+        try m.decodeFromTextArray(textDecoder: &subDecoder)
+        value = m
+        return true
+    }
+    
+    mutating func decodeRepeatedMessageField<M: ProtobufMessage>(fieldType: M.Type, value: inout [M]) throws -> Bool {
+        var token: ProtobufTextToken
+        if let startToken = try scanner.next() {
+            switch startToken {
+            case .endArray: return true // Empty array case
+            default: token = startToken
+            }
+        } else {
+            throw ProtobufDecodingError.truncatedInput
+        }
+        
+        while true {
+            switch token {
+            case .beginObject:
+                var message = M()
+                var subDecoder = ProtobufTextDecoder(scanner: scanner)
+                try message.decodeFromTextObject(textDecoder: &subDecoder)
+                value.append(message)
+            case .boolean(_), .string(_), .number(_):
+                var message = M()
+                try message.decodeFromTextToken(token: token)
+                value.append(message)
+            case .null:
+                if let message = try M.decodeFromTextNull() {
+                    // Sometimes 'null' is a valid message value
+                    value.append(message)
+                } else {
+                    // Otherwise, null is not allowed in repeated fields
+                    throw ProtobufDecodingError.malformedJSON
+                }
+            default:
+                throw ProtobufDecodingError.malformedJSON
+            }
+            if let separatorToken = try scanner.next() {
+                switch separatorToken {
+                case .comma:
+                    if let t = try scanner.next() {
+                        token = t
+                    } else {
+                        throw ProtobufDecodingError.truncatedInput
+                    }
+                    break
+                case .endArray:
+                    return true
+                default:
+                    throw ProtobufDecodingError.malformedJSON
+                }
+            }
+        }
+    }
+    
+    mutating func decodeRepeatedGroupField<G: ProtobufMessage>(fieldType: G.Type, value: inout [G]) throws -> Bool {
```

Comment:
There's a subtle difference with regard to name capitalization: https://github.com/google/protobuf/blob/master/src/google/protobuf/text_format.cc#L445